### PR TITLE
Binary / base64 field automatic encoding and eager loading

### PR DESCRIPTION
**Pull request by joshwlewis**
This PR adds automatic support for base64 / binary content, and is supported by the changes to type signatures in https://github.com/forcedotcom/sf-fx-sdk-nodejs/pull/141. 

The additional functionality has two parts:

- Population of the `binaryFields` property by eager-loading binary content during queries. 
- Automatic base64 encoding of the values in `binaryFields` when performing upstream record modification.

An example function leveraging this new functionality might look like this:

```ts
import { InvocationEvent, Context, Logger, Record, RecordModificationResult } from "sf-fx-sdk-nodejs";
import sharp from 'sharp';

/**
 * thumber queries an org for images matching a title, then creates a 50px png
 * thumbnail of each image, and uploads each thumbnail as a new image.
 */

export default async function thumber(event: InvocationEvent<any>, context: Context, logger: Logger):
    Promise<PromiseSettledResult<RecordModificationResult>[]> {

    const queryResult = await context.org.dataApi
        .query(`SELECT Id, Title, VersionData FROM ContentVersion WHERE Title = '${event.data.title}' AND FileExtension IN ('svg', 'png', 'jpg')`);

    return await Promise.allSettled(queryResult.records.map((record) => createThumb(context, record)));
}

async function createThumb(context: Context, record: Record): Promise<RecordModificationResult> {
  const resizedData = await sharp(record.binaryFields.versiondata)
    .resize(50)
    .png()
    .toBuffer();
  return await context.org.dataApi.create({
    type: "ContentVersion",
    binaryFields: {
      VersionData: resizedData,
    },
    fields: {
      Title: `${record.fields.title}-thumbnail`,
      PathOnClient: `${record.fields.title}-thumnail.png`,
      Description: `Thumbnail generated from file ${record.fields.id}`,
    }
  });
}
```

GUS-W-11543492

### CODE REVIEW

File: test/sdk/data-api.ts [hunk 1]
```diff
@@ -465,12 +502,30 @@ describe("DataApi Class", async () => {
         }
       });
     });
+
+    describe("with binary / base64 field data", async () => {
+      it("encodes binaryFields data", async () => {
```
Comment (colincasey) [inline test/sdk/data-api.ts:507]: maybe a test for verifying that a binary field can be cleared should also be added?

File: src/sdk/data-api.ts [hunk 2]
```diff
@@ -139,16 +143,7 @@ export class DataApiImpl implements DataApi {
...
-      // our SDK definition requires customers to provide "id". Customers that are not using TS might also
-      // pass other casings for the "id" field ("iD", "ID"). Any other fields are passed to the Salesforce API untouched.
-      const fields = { Id: null };
-      Object.keys(recordUpdate.fields).forEach((key) => {
-        const value = recordUpdate.fields[key];
-        const targetKey = key.toLowerCase() === "id" ? "Id" : key;
-
-        fields[targetKey] = value;
-      });
+      const fields = { Id: null, ...buildUploadFields(recordUpdate) };
```
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts:146]: The `{ Id: null }` bit is there to satisfy the type constraints of `RecordForUpdate`, which wants an `Id`. `buildUploadFields` should add the `Id`, but the TS compiler can't prove it, I guess. I'll see if I can find a better approach.

File: src/sdk/data-api.ts [hunk 3]
```diff
@@ -29,6 +30,7 @@ const pkgPath = join(
 );
 const pkg = readFileSync(pkgPath, "utf8");
 const ClientVersion = JSON.parse(pkg).version;
+const knownBinaryFields = { ContentVersion: ["VersionData"] };
```
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts:33]: I'm not sure what you mean. The data we get back from the API gives no real hint as to which fields might be binary. Seems like you have to check the docs to know. We've got a static list of known binary fields on the Java implementation too.

File: src/sdk/data-api.ts [hunk 4]
```diff
@@ -283,3 +278,57 @@ export class DataApiImpl implements DataApi {
...
+  const fields = createCaseInsensitiveMap(data);
+  delete fields["attributes"];
+
+  // For any known binaryFields, eagerly fetch the data from the specified
+  // endpoint.
+  const binaryFields = {};
+  if (type in knownBinaryFields) {
+    for (const binFieldName of knownBinaryFields[type]) {
+      if (fields[binFieldName]) {
+        const body: string = await conn.request(fields[binFieldName]);
```
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts:293]: Hmm, good point. I should probably catch and rethrow with some human readable context.

File: src/sdk/data-api.ts [hunk 5]
```diff
@@ -283,3 +278,57 @@ export class DataApiImpl implements DataApi {
...
+}
+
+function buildUploadFields(record: Record): { [key: string]: unknown } {
+  const fields = {};
+  // Automatically base64 encode any known binaryFields without overwriting existing fields.
+  if (record.type in knownBinaryFields) {
+    for (const binFieldName of knownBinaryFields[record.type]) {
+      if (
+        Buffer.isBuffer(record.binaryFields[binFieldName]) &&
+        (record.fields[binFieldName] === undefined ||
```
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts:316]: I wanted to make sure we weren't overwriting any data the user already put there. I don't feel strongly about this though. Should we drop this condition?

File: src/utils/maps.ts [hunk 6]
```diff
@@ -19,18 +17,6 @@ const mapHandler = {
...
-
-  delete fields["attributes"];
-
-  return {
-    type,
-    fields,
-  };
-}
-
 export function createCaseInsensitiveMap(map: any): any {
```
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/utils/maps.ts:20]: Yeah, I'm not loving that either. I've kind of followed the existing paradigm for what we already had. If we don't/can't support  case-insensitive fields in a bi-directional manner, we should probably abandon it. Maybe that's something we can add once we have a versioning story.

File: src/sdk/data-api.ts [hunk 7]
```diff
@@ -283,3 +278,57 @@ export class DataApiImpl implements DataApi {
...
+          record.fields[binFieldName] === null)
+      ) {
+        fields[binFieldName] =
```
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts:326]: I think we've checked all the potential null's.

File: test/sdk/data-api.ts (hunk 1 above)
Comment (joshwlewis) [reply test/sdk/data-api.ts:530]: `VersionData` is required, so there's no scenario where a dev would clear the field upstream. I did add a test to ensure that if `binaryFields` wasn't provided or was just an empty object, that we could still update the other fields. It's a good thing I added the test, because it did uncover a potential bug!

File: src/sdk/data-api.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (colincasey) [reply src/sdk/data-api.ts:33]: if you describe the `ContentVersion` field for `VersionData` it does indicate that `type=base64` and `soapType=xsd:base64Binary`. 

```sh
sfdx force:schema:sobject:describe -s ContentVersion | jq '.fields[] | select(.name == "VersionData")'
```

And the description of the `base64` type [here](https://developer.salesforce.com/docs/atlas.en-us.236.0.object_reference.meta/object_reference/primitive_data_types.htm) states:

> Base 64-encoded binary data. Fields of this type are used for storing binary files in Attachment records, Document records, and Scontrol records. In these objects, the Body or Binary field contains the (base64 encoded) data, while the BodyLength field defines the length of the data in the Body or Binary field. In the Document object, you can specify a URL to the document instead of storing the document directly in the record.

which makes me wonder if the client should attempt to fetch and cache sObject metadata to use for driving some of this logic instead of having a hardcoded list.

File: src/sdk/data-api.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (colincasey) [reply src/sdk/data-api.ts]: maybe using `in` for [type narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing)?  e.g.;

```typescript
const uploadFields = buildUploadFields(recordUpdate)
if ('Id' in uploadFields) {
  return uploadFields
}
throw new Error('hey, did you forget the Id field?')
```

File: src/sdk/data-api.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (colincasey) [reply src/sdk/data-api.ts:323]: i'm not sure, i just wanted to understand the reasoning 😅 

let's say i have a `ContentVersion` record that i want to update and i call:

```js
dataApi.update({
  type: 'ContentVersion',
  fields: {
    VersionData: 'this should not be set as a field because it is binary data'
  },
  binaryFields: {
    VersionData: bufferOfBinaryData
  }
})
```

the code, if i'm understanding it correctly, would ignore the `VersionData` value in `binaryFields` and leave the one from `fields`?  if that's true wouldn't that be an error?  or am i just misunderstanding this case?

File: src/sdk/data-api.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts:33]: The java implementation is shipping with a static list too: 
https://github.com/forcedotcom/sf-fx-runtime-java/blob/51b288979acd40d560023a33a854010dd2bd7d20/sf-fx-runtime-java-sdk-impl-v1.1/src/main/java/com/salesforce/functions/jvm/runtime/sdk/BinaryFieldUtil.java#L22

Though, I do like the programatic approach for generating these. @Malax what do you think?

[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts:33]: @colincasey and I decided this would be a nice future improvement, and could also potentially add functionality where we have fully typed sObjects in later versions.

**Review by colincasey — APPROVED** [review]

File: src/sdk/data-api.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (joshwlewis) [reply src/sdk/data-api.ts]: We're now throwing an error in this scenario instead: https://github.com/forcedotcom/sf-fx-runtime-nodejs/pull/400/commits/fb99693d129a9748e85feb3413806ff5b03de7a1
